apps/internal/base/base.go

```diff
@@ -257,6 +268,72 @@ func (b Client) AcquireTokenByAuthCode(ctx context.Context, authCodeParams Acqui
 	return b.AuthResultFromToken(ctx, authParams, token, true)
 }
 
+// AcquireTokenByInteractive initiates the interactive auth flow.
+func (b Client) AcquireTokenByInteractive(ctx context.Context, params AcquireTokenInteractiveParameters) (AuthResult, error) {
+	// the code verifier is a random 32-byte sequence that's been base-64 encoded without padding.
+	// it's used to prevent MitM attacks during auth code flow, see https://tools.ietf.org/html/rfc7636
+	cv, hash, err := codeVerifier()
+	if err != nil {
+		return AuthResult{}, err
+	}
+	authParams := b.AuthParams // This is a copy, as we dont' have a pointer receiver and .AuthParams is not a pointer.
+	authParams.Scopes = params.Scopes
+	authParams.AuthorizationType = authority.ATInteractive
+	authParams.CodeChallenge = hash
+	authParams.CodeChallengeMethod = "S256"
+	authParams.State = uuid.New().String()
+	authParams.Prompt = "select_account"
+	res, err := b.browserLogin(ctx, authParams)
+	if err != nil {
+		return AuthResult{}, err
+	}
+	authParams.Redirecturi = res.redirectURI
+
+	req, err := accesstokens.NewCodeChallengeRequest(authParams, accesstokens.ATPublic, nil, res.authCode, cv)
```

Comment:
I think this function is a little misleading. It mentions new codeChallengeRequest when actually we are sending a code-verifier to the token endpoint. Code challenge is something we send in the request to authcode endpoint. That would also mean changes [here](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/apps/internal/oauth/ops/accesstokens/accesstokens.go#L189) and [here](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/apps/internal/oauth/ops/accesstokens/accesstokens.go#L218)

Reply:
Not sure I follow.  You still have to send the code verifier when requesting the token (the code verifier hash is sent in the challenge when requesting the auth code, the code verifier is sent with the token request).

Reply:
Yes, what I mean is we are sending a CodeVerifier in this function named  `NewCodeChallengeRequest` along with an authCode to get back an accessToken. 
However, the name `NewCodeChallengeRequest` is a little bit odd because :
1. It is not a request to get a CodeChallenge
2. It is not requesting something using a CodeChallenge
3. And then in this [line](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/apps/internal/oauth/ops/accesstokens/accesstokens.go#L218) we set a code_verifier with a value of something named `CodeChallenge` which is again misleading.

Let me know if this makes sense, happy to sync offline to elaborate.

Reply:
I see your point now.  Agreed the name isn't ideal, perhaps something like `NewRequestWithCodeVerifier()` or something along those lines.  It's not public surface area so we can change it now or at any time.

Reply:
`NewRequestwithCodeVerifier()` is good. That would probably also mean changes [here](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/apps/internal/oauth/ops/accesstokens/accesstokens.go#L189) and [here](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/apps/internal/oauth/ops/accesstokens/accesstokens.go#L218). But its fine if these are not changed in this PR. I can make these changes after this is merged in.

Reply:
We can change it but we should still be mindful of the terminology.  While both legs of the flow send it via the `code_challenge` param, in the first leg of the flow it's called the challenge which is where we send the S256 hash of the code verifier.  The second leg of the flow sends the actual code verifier (same query param, different values).
So for APIs used in the first leg we should still call the param a `challenge`, and only in the second leg of the flow we should call it `codeVerifier`.

Reply:
I started to do this but then noticed some other oddities with auth code flow which need to be fixed.  I've opened https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/155 to track refactoring all of this.

Reply:
Yeah that makes sense! This needs to be refactored.